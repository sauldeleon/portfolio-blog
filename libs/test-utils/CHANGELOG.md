# Changelog Test utils

## [0.2.3](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.2...test-utils-0.2.3) (2024-01-08)

## [0.2.2](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.1...test-utils-0.2.2) (2023-09-26)

## [0.2.1](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.0...test-utils-0.2.1) (2023-09-14)

# [0.2.0](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.1.0...test-utils-0.2.0) (2023-09-05)

### Bug Fixes

- reduced duplicated lines ([f51168a](https://github.com/sauldeleon/portfolio-blog/commit/f51168aefe025a5a65ffdb2dc22f7a16b863406e))
- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))

# 0.1.0 (2023-09-02)

### Bug Fixes

- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- add contact initial page ([fea862b](https://github.com/sauldeleon/portfolio-blog/commit/fea862b3507268013149050ae3f05fa517e7b3d9))
- add homepage ([60523aa](https://github.com/sauldeleon/portfolio-blog/commit/60523aa31ec088d318745ef0c757b4cbe656f0d4))
- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))
- improved a11y and styles ([873f656](https://github.com/sauldeleon/portfolio-blog/commit/873f656587f50df8e29ad526fdc68fea0b2c9784))
- toggle language button ([afd97e9](https://github.com/sauldeleon/portfolio-blog/commit/afd97e95beac0f70ed643c753e71db2ca014054b))
