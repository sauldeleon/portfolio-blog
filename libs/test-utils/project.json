{
  "name": "test-utils",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/test-utils/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "version": {
      "executor": "@jscutlery/semver:version",
      "options": {
        "changelogHeader": "# Changelog Test utils \n",
        "baseBranch": "main",
        "commitMessageFormat": "chore(${projectName}): release: v${version}",
        "push": false
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/test-utils/**/*.{ts,tsx,js,jsx}"]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/test-utils/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  }
}
