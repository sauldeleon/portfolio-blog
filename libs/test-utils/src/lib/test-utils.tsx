import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { Matcher } from '@testing-library/dom/types/matches'
import {
  RenderOptions,
  RenderResult,
  queryAllByAttribute as _queryAllByAttribute,
  queryByAttribute as _queryByAttribute,
  act,
  render,
  screen,
} from '@testing-library/react'
import { RouterContext } from 'next/dist/shared/lib/router-context'
import Router, { NextRouter } from 'next/router'
import { ReactNode, useEffect, useState } from 'react'
import { ThemeProvider } from 'styled-components'

import { LanguageContextProvider } from '@sdlgr/i18n-config'
import { mainTheme } from '@sdlgr/main-theme'

export type MatcherCustomFnParams = {
  content: string
  element: Element | null
}

/* istanbul ignore next */
export const renderWithTheme = (children: ReactNode): RenderResult => {
  return render(<ThemeProvider theme={mainTheme}>{children}</ThemeProvider>)
}

/* istanbul ignore next */
export function RenderProviders({ children }: { children: ReactNode }) {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

  return (
    <QueryClientProvider client={queryClient}>
      <LanguageContextProvider value={{ language: 'en' }}>
        <ThemeProvider theme={mainTheme}>{children}</ThemeProvider>
      </LanguageContextProvider>
    </QueryClientProvider>
  )
}

/* istanbul ignore next */
export const renderApp = (
  children: ReactNode,
  options?: Omit<RenderOptions, 'queries'>
) => {
  return render(<RenderProviders>{children}</RenderProviders>, options)
}

/**
 * Generic function to get element by attribute and value
 * @example queryByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function queryByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example queryByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function queryAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Generic function to get element by attribute and value
 * @example getByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function getByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example getByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function getAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Shortcut version for queryByAttribute, but specific for id
 * @param {HTMLElement} container
 * @param {string} id
 * @return {any}
 */
export function getById(container: HTMLElement, id: string) {
  return queryByAttribute('id')(container, id) as Element
}

/**
 * Shortcut version for queryByAttribute, but specific for class name
 * @param {HTMLElement} container
 * @param {string} className
 * @return {any}
 */
export function getByClassName(container: HTMLElement, className: string) {
  return queryByAttribute('class')(container, className) as Element
}

function matcherCustomFn(
  tag: string,
  matcher: Matcher,
  { element, content }: MatcherCustomFnParams
) {
  const isTag = element?.tagName.toUpperCase() === tag.toUpperCase()
  let matching = false
  if (isTag) {
    matching = content === matcher
    /* istanbul ignore else*/
    if (typeof matcher === 'function') {
      matching = matcher(content, element)
    } else if (matcher instanceof RegExp) {
      matching = matcher.test(content)
    }
  }
  return matching && isTag
}

/**
 * Get 1 element by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getByTagAndText(tag: string, matcher: Matcher) {
  return screen.getByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Get multiple elements by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getAllByTagAndText(tag: string, matcher: Matcher) {
  return screen.getAllByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Mock Math.random() to return a specific value
 * @param {number} value
 */
export function mockRandom(value = 0.1234) {
  const mockMath = Object.create(global.Math)
  mockMath.random = () => value
  global.Math = mockMath
}

/**
 * Hook to use in test to ensure "last" render is called
 */
export function useWithRefreshInTest() {
  const [withRefreshInTest, setWithRefreshInTest] = useState(false)

  useEffect(() => {
    // Change state to "rerender" the component
    setWithRefreshInTest(true)
  }, [])

  return withRefreshInTest
}

export { render } from '@testing-library/react'

/* istanbul ignore next */
export const awaitAllAsync = async () =>
  act(() => new Promise((resolve) => setTimeout(resolve, 0)))

/* istanbul ignore next */
export function renderWithNextRouter(
  children: ReactNode,
  options?: Partial<NextRouter>
) {
  return renderWithTheme(
    <RouterContext.Provider value={{ ...Router, ...options }}>
      {children}
    </RouterContext.Provider>
  )
}
