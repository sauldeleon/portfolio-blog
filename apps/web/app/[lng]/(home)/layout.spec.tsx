import { screen } from '@testing-library/react'

import { renderApp } from '@sdlgr/test-utils'

import HomeLayout from './layout.next'

describe('/[lng]/(home) - HomeLayout', () => {
  it('should render', async () => {
    const { baseElement } = renderApp(<HomeLayout>test</HomeLayout>)
    await screen.findByRole('heading', { level: 1 })
    expect(screen.getByText('test')).toBeInTheDocument()
    expect(baseElement).toMatchSnapshot()
  })
})
