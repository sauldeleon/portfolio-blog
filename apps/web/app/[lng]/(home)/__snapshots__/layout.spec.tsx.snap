// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/[lng]/(home) - HomeLayout should render 1`] = `
.c0 {
  font-family: 'Roboto mono',monospace;
  font-weight: 400;
  color: #FBFBFB;
  font-size: 1.5rem;
  line-height: normal;
  margin: 0;
}

.c5 {
  font-family: 'Roboto mono',monospace;
  font-weight: 400;
  color: #FBFBFB;
  font-size: 0.9rem;
  line-height: normal;
  white-space: pre-line;
}

.c5 ul,
.c5 ol {
  margin-left: 40px;
  margin-bottom: 10px;
}

.c5 ul {
  list-style: outside disc;
}

.c5 ol {
  list-style: decimal;
}

.c5 li {
  padding-left: 10px;
}

.c6 {
  position: relative;
  position: relative;
}

.c6::after {
  content: '';
  position: absolute;
  width: 100%;
  height: 2px;
  bottom: 0;
  left: 0;
  background-color: #FBFBFB;
}

.c6::after {
  transform: scaleX(0);
  transform-origin: bottom right;
  transition: transform 0.25s ease-out;
}

.c6:hover::after {
  transform: scaleX(1);
  transform-origin: bottom left;
}

.c6:hover,
.c6:active,
.c6:focus {
  text-decoration: none;
}

.c13 {
  margin-left: 8px;
  position: relative;
}

.c13::after {
  content: '';
  position: absolute;
  width: 100%;
  height: 2px;
  bottom: 0;
  left: 0;
  background-color: #FBFBFB;
}

.c13::after {
  transform: scaleX(0);
  transform-origin: bottom right;
  transition: transform 0.5s ease-out;
}

.c7 {
  outline: none;
  cursor: pointer;
  display: flex;
  align-items: center;
  max-height: 42px;
}

.c7::after {
  content: none;
}

.c7:hover circle.loading {
  stroke: white;
  stroke-width: 6px;
  stroke-dasharray: 110;
  stroke-dashoffset: 82;
  animation: kVEmCK 0.5s steps(4, jump-none) forwards;
  transform: rotate(-90deg);
  transform-origin: center;
}

.c7:hover .c12::after {
  transform: scaleX(1);
  transform-origin: bottom left;
}

.c9 {
  position: relative;
  width: 42px;
  height: 42px;
}

.c10 {
  position: absolute;
  transform: translate(-50%, -50%);
  top: 50%;
  left: 50%;
}

.c11 {
  position: absolute;
  transform: translate(-50%, -50%);
  top: 50%;
  left: 50%;
  display: flex;
}

.c1 {
  text-align: center;
}

.c2 {
  margin-top: 30px;
}

.c8 {
  margin-top: 40px;
  margin-bottom: 30px;
}

.c3 {
  text-align: center;
  max-width: 1180px;
  padding: 0 34px;
}

.c3 .c4 {
  margin-top: 20px;
}

@media (min-width: 768px) {
  .c1 {
    font-family: 'Roboto mono',monospace;
    font-weight: 400;
    color: #FBFBFB;
    font-size: 3rem;
    line-height: normal;
  }
}

@media (min-width: 1024px) {
  .c1 {
    margin-bottom: 10px;
  }
}

@media (min-width: 768px) {
  .c2 {
    margin-bottom: 40px;
    font-family: 'Roboto mono',monospace;
    font-weight: 400;
    color: #FBFBFB;
    font-size: 3rem;
    line-height: normal;
  }
}

@media (min-width: 768px) {
  .c8 {
    margin-bottom: 50px;
  }
}

@media (min-width: 768px) {
  .c3 .c4 {
    font-family: 'Roboto mono',monospace;
    font-weight: 400;
    color: #FBFBFB;
    font-size: 1rem;
    line-height: normal;
  }
}

<body>
  <div>
    <h1
      class="c0 c1"
    >
      Saúl de León Guerrero
    </h1>
    test
    <h2
      class="c0 c2"
    >
      About me
    </h2>
    <section
      class="c3"
    >
      <p
        class="c4 c5"
      >
        Caffeine and music alchemist, turning beverages and tunes into web applications. I have a soft spot for the xERN stack, transforming designs into responsive webs. I'm fluent in HTML5, CSS3, and (Java|Type)Script.
      </p>
      <p
        class="c4 c5"
      >
        I'm not just a one-trick pony. While I'm more focused on FrontEnd, I also have a blast developing primarily with Node.js, but I'm learning new technologies and having experience with other technologies too.
      </p>
      <p
        class="c4 c5"
      >
        I do love staying up-to-date with the latest trends, and my commitment is to deliver solutions that can be improved and maintained without suffering pain (if possible).
      </p>
    </section>
    <a
      class="c6 c7 c8"
      href="/experience"
    >
      <div
        class="c9"
      >
        <svg
          class="c10"
          height="42"
          width="42"
        >
          circleButton.svg
        </svg>
        <div
          class="c11"
        >
          <svg>
            arrowRight.svg
          </svg>
        </div>
      </div>
      <p
        class="c4 c12 c5 c13"
      >
        Experience
      </p>
    </a>
  </div>
</body>
`;
